### PR TITLE
feat(cli): collect analytics

### PR DESCRIPTION
**Pull request by hasanaburayyan**
Currently this PR introduces some recorded analytics that can be opted out of using `WING_DISABLE_ANALYTICS` env var. 

Here are some example analytics:
- User running from non git repo
image

- User running from git repo
	- Note `anonymous_repo_id` has been added
image

- Automation running in CI
	- Note `anonymous_repo_id` is added
	- Note `anonymous_id` says 'ci_run' 

image

### Implementation Details

In order to ensure we are not slowing down the user experience when running any CLI command, the process for collecting metrics works in 3 main components. Similar to how other CLI tools I have come across collect metrics. 
- Collectors: Which is responsible for gathering up all the analytics we care about during runtime of the CLI.
- Storage: The analytics gathered by the collector are stored on disk. The storage will house the events until the exporter is able to report the events back to [segment](https://segment.com/)
- Exporter: To avoid any long delays in the CLI commands, the exporter is run as a detached process when the CLI command is complete.

Right now the best way I am seeing to collect cli command specific data is by using Commander hooks around all the commands we want to collect analytics from (which right now I have as all of them)

## TODO:
- [x] Add notification on first run about analytic collection
- [x] Docs
	- [x] 	Document all data being collected
	- [x] Document opt-out policy
- [x] Determine if CLI called within CI
- [x] Collect git metrics (if exists)
- [ ] Legal review
- [ ] Github disclaimer
- [x] Setup privacy email
- [x] Use tmp dir for analytics storage
- [x] Consider best way to manage WriteKey for segment

Closes: https://github.com/winglang/wing/issues/2928
Closes: https://github.com/winglang/wing/issues/2974

## Checklist

- [x] Title matches [Winglang's style guide](https://docs.winglang.io/contributing/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [x] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: apps/wing/src/analytics/storage.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  return config.anonymousId;
+}
+
+export function loadAnalyticsConfig(): AnalyticsConfig {
+  try {
+    const fileContents = readFileSync(WING_ANALYTICS_CONFIG_FILE, 'utf-8');
+    return JSON.parse(fileContents) as AnalyticsConfig;
+  } catch (error) {
+    // Possible issue reading analytics config file, first determine if file exists yet
+    if (existsSync(WING_ANALYTICS_CONFIG_FILE)) {
```
Comment (MarkMcCulloh) [inline apps/wing/src/analytics/storage.ts:55]: nit: Instead of checking for existence, check out the error codes from the error to see if it's ENOENT or EACCES.

File: apps/wing/src/analytics/disclaimer.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import chalk from "chalk";
+import { loadAnalyticsConfig, saveAnalyticsConfig } from "./storage";
+
+const disclaimer = `
+🧪 This is an early pre-release of the Wing Programming Language (aka "alpha").
```
Comment (eladb) [inline apps/wing/src/analytics/disclaimer.ts:5]:
```suggestion
🧪 This is an early pre-release of the Wing Programming Language.
```

File: apps/wing/src/analytics/event.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import { CLIData } from "./collectors/cli-collector";
+import { NodeData } from "./collectors/node-collector";
+import { OSData } from "./collectors/os-collector";
+
+export interface AnalyticEvent {
+  event: string;
+  timestamp?: string;
+  anonymousId?: string;
+  properties: {
+    inCI: boolean;
```
Comment (eladb) [inline apps/wing/src/analytics/event.ts:11]: do we really need both `inCI` and `ci`. We can assume that if `ci` has contents we are in a CI context, no?

File: apps/wing/src/analytics/scripts/detached-export.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { getAnonymousId, loadAnalyticsEvent } from '../storage';
+import * as fs from 'fs';
+import Analytics from '@segment/analytics-node';
+import { getWingAnalyticsCollectionConfig } from '../collect';
+
+// When this file is run as a child process, it will be passed the path 
+// to the analytics report file
+const filePath = process.argv[2];
+
```
Comment (eladb) [inline apps/wing/src/analytics/scripts/detached-export.ts:9]: error is argv[2] is not supplied

File: apps/wing/src/analytics/scripts/detached-export.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      properties: event.properties,
+    });
+  }
+
+  // Keeps the file around for debugging purposes
+  if (!analyticsConfig.debug) {
+    fs.unlinkSync(filePath);
+  }
+} catch(err) {
+  // TODO: add mechanism to retry
```
Comment (eladb) [inline apps/wing/src/analytics/scripts/detached-export.ts:29]: i am assuming if we fail, we will simply send the file again next time, no? that seems like a reasonable retry

File: apps/wing/src/analytics/storage.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ * @param event the event to store
+ * @returns string the file path of the stored analytic
+ */
+export function storeAnalyticEvent(event: AnalyticEvent): string {
+  const eventId = uuidv4();
+  const tmpdir = path.join(os.tmpdir(), "wing-analytics");
+  
+  if (!existsSync(tmpdir)) {
+    mkdirSync(tmpdir);
+  }
```
[the remark this answers is not in the document]
Comment (eladb) [reply apps/wing/src/analytics/storage.ts:29]: i think we *never* want to fail if something doesn't work with our analytics. probably need to surround everything with try/catch like you did in the `detached-export` script

**Comment by hasanaburayyan** [comment]
@eladb and @MarkMcCulloh thanks for starting to take a look at this one. Im going to add the `do-not-merge` I forgot I made this not a draft and still have TODOs 😁 but love the feedback keep it coming

File: apps/wing/src/cli.ts [hunk 7]
```diff
@@ -7,7 +7,16 @@ import { satisfies } from "compare-versions";
...
-const PACKAGE_VERSION = require("../package.json").version as string;
+import { collectCommandAnalytics, getWingAnalyticsCollectionConfig } from "./analytics/collect";
+import { exportAnalytics } from "./analytics/export";
+import { optionallyDisplayDisclaimer } from "./analytics/disclaimer";
+
+
+export const PACKAGE_VERSION = require("../package.json").version as string;
+
+const ANALYTICS_COLLECTION_CONFIG = getWingAnalyticsCollectionConfig(); 
+let ANALYTICS_EXPORT_FILE: string | undefined = undefined;
```
Comment (eladb) [inline apps/wing/src/cli.ts:18]: ALL_CAPS usually represent constants

File: docs/docs/07-examples/14-wing-analytics.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
+---
+title: Wing Cli Analytics
+id: wing-analytics
+keywords: [Analytics, Telemetry]
+---
+
+## Wing Anonymous Analytics
+
+The Wing cli collects anonymous usage data in order to help us improve the cli experience. The intent is to collect data that will help us understand how the cli is being used throughout the community, with hopes that this will help us to build the right features and improve the right areas.
```
Comment (eladb) [inline docs/docs/07-examples/14-wing-analytics.md:9]:
```suggestion
The Wing CLI collects anonymous usage data in order to help us improve. The intent is to collect data that will help us understand how the CLI is being used throughout the community, with hopes that this will help us to build the right features and improve the right areas.
```

File: docs/docs/07-examples/14-wing-analytics.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+You may not feel comfortable sharing this data with us, and that is ok. The gathering of these analytics is completely optional to use the Wing cli, and can be opted out of at any time.
```
Comment (eladb) [inline docs/docs/07-examples/14-wing-analytics.md:11]:
```suggestion
You may not feel comfortable sharing this data with us, and that is ok. The gathering of these analytics is completely optional to use the Wing CLI, and can be opted out of at any time.
```

File: docs/docs/07-examples/14-wing-analytics.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+```sh
+wing compile -t sim app.w --no-analytics
+```
+
+### What data is collected
+
+Each run of the Wing cli collects information about various aspects of the environment it is running in. Below are the analytics we gather. As well
+you can always head over to our Github and view the code that collects these metrics, as the code is open source.
+
+- The Wing cli version
```
Comment (eladb) [inline docs/docs/07-examples/14-wing-analytics.md:26]:
```suggestion
- The Wing CLI version
```

File: apps/wing/src/analytics/scripts/detached-export.ts (hunk 5 above)
Comment (hasanaburayyan) [reply apps/wing/src/analytics/scripts/detached-export.ts:29]: currently in order to avoid a race condition (if multiple wing commands run in parallel)  the detached script is called with a specific analytic file uuid, so there would need to be some sort of watch or failure log or something for the subsequent runs to know to try exporting the failed metric.

File: docs/docs/07-examples/14-wing-analytics.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+- The command information, including the command name, arguments, and flags
+- Operating system information, including the architecture, version, and platform
+- Node js version
```
Comment (eladb) [inline docs/docs/07-examples/14-wing-analytics.md:29]:
```suggestion
- Node.js version
```

File: docs/docs/07-examples/14-wing-analytics.md [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+- CI environment name (if applicable)
```
Comment (eladb) [inline docs/docs/07-examples/14-wing-analytics.md:30]: What about Wing SDK version? Console version?

File: docs/docs/07-examples/14-wing-analytics.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
+---
```
Comment (eladb) [inline docs/docs/07-examples/14-wing-analytics.md:1]: Why is this under 07-examples?

File: docs/docs/07-examples/14-wing-analytics.md [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+title: Wing Cli Analytics
```
Comment (eladb) [inline docs/docs/07-examples/14-wing-analytics.md:2]:
```suggestion
title: Analytics
```

File: docs/docs/07-examples/14-wing-analytics.md [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+id: wing-analytics
+keywords: [Analytics, Telemetry]
+---
+
+## Wing Anonymous Analytics
```
Comment (eladb) [inline docs/docs/07-examples/14-wing-analytics.md:7]:
```suggestion
## Analytics Collection
```

File: docs/docs/07-examples/14-wing-analytics.md [hunk 16]
```diff
@@ -0,0 +1,30 @@
+---
+title: Wing Cli Analytics
+id: wing-analytics
```
Comment (eladb) [inline docs/docs/07-examples/14-wing-analytics.md:3]:
```suggestion
id: analytics
```

File: apps/wing/src/analytics/event.ts (hunk 3 above)
Comment (hasanaburayyan) [reply apps/wing/src/analytics/event.ts:11]: I went back and forth on this. In my mind it would depend on what we are using downstream to query/visualize these analytics. I assume you are right, but I wonder if having this easy true/false field would make some query easier downstream. 🤷 Happy to take it out, I wonder if @ekeren has any thoughts here?

File: docs/docs/07-examples/14-wing-analytics.md (hunk 13 above)
Comment (hasanaburayyan) [reply docs/docs/07-examples/14-wing-analytics.md:1]: Just a placeholder for now, wanted to confirm where this was actually going to go.

File: apps/wing/src/cli.ts [hunk 17]
```diff
@@ -83,6 +126,7 @@ async function main() {
     )
     .option("-p, --plugins [plugin...]", "Compiler plugins")
     .hook("preAction", progressHook)
+    .hook("preAction", collectAnalyticsHook)
```
[the remark this answers is not in the document]
Comment (hasanaburayyan) [reply apps/wing/src/cli.ts:129]: let me try that again, I could not get "preSubcommand" to work last week (iirc I was not getting some command information I expected) but perhaps I was goofing something up. I would much rather have it once and done.

[the remark this answers is not in the document]
Comment (hasanaburayyan) [reply apps/wing/src/cli.ts:129]: Ah I see I was trying to parse the root command rather than the sub command when using that hook, which is why I thought data was missing. This works like a charm good call

File: apps/wing/src/cli.ts [hunk 18]
```diff
@@ -53,6 +66,36 @@ async function main() {
...
     });
 
+
+    async function collectAnalyticsHook(cmd: Command) {
+      if (!ANALYTICS_COLLECTION_CONFIG.collect) { return; }
+  
+      // Fail silently if collection fails
+      try {
+        optionallyDisplayDisclaimer();
+        ANALYTICS_EXPORT_FILE = await collectCommandAnalytics(cmd);
```
[the remark this answers is not in the document]
Comment (hasanaburayyan) [reply apps/wing/src/cli.ts]: ahh I see, yea that voodoo works I had not thought about it. 😅

File: apps/wing/src/analytics/scripts/detached-export.ts (hunk 5 above)
Comment (MarkMcCulloh) [reply apps/wing/src/analytics/scripts/detached-export.ts:29]: imo I don't think we should burden the user's system/io with retries. Metrics are great but let's not be greedy, especially since our metrics are opt-out. 
At most we should retry once in this same process (which luckily doesn't block anything) and if it doesn't work then just consider the data lost.

File: apps/wing/src/analytics/storage.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (MarkMcCulloh) [reply apps/wing/src/analytics/storage.ts:29]: For this particular chunk, the exception is actually used for logic rather than just handling failures. But yeah I would agree this whole thing also needs a try/catch anyways.

File: apps/wing/src/analytics/collectors/ci-collector.ts [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+  CIRCLECI = "circleci",
+  BITBUCKET = "bitbucket",
+  AZURE_DEVOPS = "azure_devops",
+  TEAMCITY = "teamcity",
+  AWS_CODEBUILD = "aws_codebuild",
+  NONE = "none",
+}
+
+export class CICollector extends Collector {
+
```
Comment (eladb) [inline apps/wing/src/analytics/collectors/ci-collector.ts:20]: Nit: This class feels a bit over designed...

I'd just write it like so:

```js
// list of supported CI environment variables
const ENV = [
  "GITHUB_ACTIONS",
  "GITLAB_CI",
  // ...
};

export class CICollector extends Collector {
  async collect(): Promise<CIData | undefined> {
    for (const e of ENV) {
      if (e in process.env) {
        return { name: e };
      }
    }

    return undefined;
  }
}
```

I don't see a real need to mangle the env name in the `name` and also, we can update the contract of the `Collector` to allow `undefined` to be returned, indicating this collection is empty.

(this requires a change to `collect.ts` to not call `canCollect()`.

**Comment by ainvoner** [comment]
@hasanaburayyan wing and wing console should use the same analytics anonymous id.
In order for the console to have access to the analytics module, it should be developed as a library and not as part of the wing app (perhaps only the anonId creation and getter?).
can you please refactor your solution to support it?
basically the console needs only the `getAnonymousId()`

File: apps/wing/src/analytics/collectors/ci-collector.ts (hunk 19 above)
Comment (hasanaburayyan) [reply apps/wing/src/analytics/collectors/ci-collector.ts:20]: thats fair, yea this originally was a lot more complicated when I had it designed to get much more information than just the CI environment name. (repo id, commit hash, etc..) Ill simplify this

**Comment by hasanaburayyan** [comment]
> @hasanaburayyan wing and wing console should use the same analytics anonymous id. In order for the console to have access to the analytics module, it should be developed as a library and not as part of the wing app (perhaps only the anonId creation and getter?). can you please refactor your solution to support it? basically the console needs only the `getAnonymousId()`

@ainvoner sure let me take a look at extracting this today

File: apps/wing/src/analytics/collectors/ci-collector.ts (hunk 19 above)
Comment (hasanaburayyan) [reply apps/wing/src/analytics/collectors/ci-collector.ts:20]: Although with mangling the name, its not always the case that an environment id is a nice indicator of what CI it is in. Like with Jenkins we only know based on JENKINS_URL which is not really a good name, but we would know its Jenkins (which could be okay) but a step worse might be azure which is identified by the environment variable `BUILD_BUILDID` I would have no idea looking at downstream analytics that is azure

Comment (eladb) [reply apps/wing/src/analytics/collectors/ci-collector.ts:20]: So you can make `ENV` a map from variable to identifier.

File: apps/wing/src/analytics/collect.ts [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+  const ciCollector = new CICollector();
+  const cliCollector = new CLICollector(cmd);
+
+  let event: AnalyticEvent = {
+    event: `wing cli:${cmd.name()}`,
+    properties: {
+      cli: await cliCollector.collect(),
+      os: await osCollector.collect(),
+      node: await nodeCollector.collect(),
+      ci: await ciCollector.collect(),
```
Comment (ekeren) [inline apps/wing/src/analytics/collect.ts:29]: We need to flatten this all things under `properties,` nested maps are not good for our analytics consumers 

Instead of:
```
properties: { 
  cli: { 
    k1: "v1", 
    k2: { 
      k3: "v2"
    }
  }
}
```
Is should be:
```
properties: { 
  cli_k1: "v1",
  cli_k2_k3: "v2
}
```

File: apps/wing/src/analytics/storage.ts [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+    const analyticsConfig: AnalyticsConfig = {
+      anonymousId: randomBytes(16).toString('hex'),
+      optOut: false,
+    }
+
+    saveAnalyticsConfig(analyticsConfig);
+
+    return analyticsConfig;
+  }
+}
```
Comment (ekeren) [inline apps/wing/src/analytics/storage.ts:75]: @hasanaburayyan , 

I was thinking that in CI we will not have this value, the result is that in every CI run the system will generate a new `anonymousId` 

One option I have is to take the git repo and run a hash function on it (e.g. md5) and using the hashed versions  as an anonymousId in the case of CI 

I this case we can't determine the real github repo, (unless we guess it and run the same hash function) 

It might be also nice to add this anonymouRepoId in all the events, we can use it to understand correlation  between anonymouId and anonymouRepoId  without any privacy issues

File: apps/wing/src/analytics/collect.ts (hunk 20 above)
Comment (ekeren) [reply apps/wing/src/analytics/collect.ts:29]: also notice that properties should be snake case (just a naming convention)

File: apps/wing/src/analytics/storage.ts (hunk 21 above)
Comment (ekeren) [reply apps/wing/src/analytics/storage.ts:75]: actually anonymouRepoId should be inside `properties` and should be called : `anonymou_repo_id`

Comment (ekeren) [reply apps/wing/src/analytics/storage.ts:75]: @ainvoner - notice that the console should also have the `anonymou_repo_id` property

File: apps/wing/src/analytics/collect.ts (hunk 20 above)
Comment (hasanaburayyan) [reply apps/wing/src/analytics/collect.ts:29]: Love the call out. I did not think about consumers querying nested maps aspect. Will fix this conventions

File: apps/wing/src/analytics/storage.ts (hunk 21 above)
Comment (ekeren) [reply apps/wing/src/analytics/storage.ts:75]: yes, I am full of typos

File: apps/wing/src/analytics/collectors/cli-collector.ts [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+    super();
+    this.cmd = cmd;
+  }
+
+  async collect(): Promise<CLIData> {
+    return {
+      target: this.cmd.opts().target,
+      options: `${JSON.stringify(this.cmd.opts())}`,
+      version: PACKAGE_VERSION,
+      wingSDKVersion: require(`@winglang/sdk/package.json`).version as string,
```
Comment (ekeren) [inline apps/wing/src/analytics/collectors/cli-collector.ts:26]:
```suggestion
      wing_sdk_version: require(`@winglang/sdk/package.json`).version as string,
```

File: apps/wing/src/analytics/collectors/cli-collector.ts [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+      wingConsoleVersion: require(`@wingconsole/app/package.json`).version as string
```
Comment (ekeren) [inline apps/wing/src/analytics/collectors/cli-collector.ts:27]:
```suggestion
      wing_console_version: require(`@wingconsole/app/package.json`).version as string
```

File: apps/wing/src/analytics/collect.ts [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+ * @returns string the file path of the stored analytic
+ */
+export async function collectCommandAnalytics(cmd: Command): Promise<string | undefined> {
+  const osCollector = new OSCollector();
+  const nodeCollector = new NodeCollector();
+  const ciCollector = new CICollector();
+  const cliCollector = new CLICollector(cmd);
+
+  let event: AnalyticEvent = {
+    event: `wing cli:${cmd.name()}`,
```
Comment (ekeren) [inline apps/wing/src/analytics/collect.ts:24]: Some naming convention change, it is <source>_<object>_<action>
in this case:
-  source is `cli`
- object is `--target` 
- action is `test/compile/etc`
 
if object (target) is empty (like in `wing docs` it should be "nil"), unless there is a default (then it should be the default) 
Also, lets remove anything in the object/action that is not normal chars [a-zA-Z] `str.replace(/[^a-zA-Z]/g, '');`

meaning it should be:
- `name: cli_sim_test`
- `name: cli_tfaws_compile`
- `name: cli_sim_compile`
- `name: cli_tfaws_compile`

File: apps/wing/src/analytics/collectors/cli-collector.ts [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+  private cmd: Command;
+
+  constructor(cmd: Command) {
+    super();
+    this.cmd = cmd;
+  }
+
+  async collect(): Promise<CLIData> {
+    return {
+      target: this.cmd.opts().target,
```
[the remark this answers is not in the document]
Comment (hasanaburayyan) [reply apps/wing/src/analytics/collectors/cli-collector.ts:23]: Yes since the commander CLI uses sim as default for commands that do not require `--target` like `test`

File: apps/wing/src/analytics/collect.ts (hunk 24 above)
Comment (hasanaburayyan) [reply apps/wing/src/analytics/collect.ts:23]: sorry not sure if I get this completely. In the above are `cli_sim_test`, `cli_tfaws_compile`, etc.. values in `event`? Also does that mean that running docs would produce `cli_nil_docs` or `cli_docs` or `cli__docs`

File: apps/wing/src/analytics/storage.ts (hunk 21 above)
Comment (hasanaburayyan) [reply apps/wing/src/analytics/storage.ts]: Also @ekeren I believe we have to keep anonymousID even if it was in CI this is because its a required value in segment. Though if its in CI I can just make the value say "see anonymous_repo_id"

File: apps/wing/src/analytics/collect.ts (hunk 24 above)
Comment (ekeren) [reply apps/wing/src/analytics/collect.ts:25]: Let's go with `cli__docs`, I think it is the best approach

Comment (hasanaburayyan) [reply apps/wing/src/analytics/collect.ts:25]: alright heres what it looks like
image

File: apps/wing/src/cli.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (hasanaburayyan) [reply apps/wing/src/cli.ts]: @MarkMcCulloh I reverted back to adding this as a "preAction" hook to each command. It appears that the lifecycle hook for "preSubcommand" does not get the options on the subcommand. The only options I can parse are the defaults.

File: apps/wing/src/analytics/collect.ts [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+  const eventName = `cli_${cmd.opts().target ?? ''}_${cmd.name()}`;
+
+  let event: AnalyticEvent = {
+    event: eventName.replace(/[^a-zA-Z_]/g, ""),
+    properties: {
+      cli: await cliCollector.collect(),
+      os: await osCollector.collect(),
+      node: await nodeCollector.collect(),
+      ci: await ciCollector.collect(),
+      anonymous_repo_id: (await gitCollector.collect())?.anonymous_repo_id
```
Comment (eladb) [inline apps/wing/src/analytics/collect.ts:33]: Why is this not just?

```suggestion
      git: await gitCollector.collect()
```

File: apps/wing/src/analytics/collectors/cli-collector.ts [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+      options: `${JSON.stringify(this.cmd.opts())}`,
+      version: PACKAGE_VERSION,
+      wing_sdk_version: require(`@winglang/sdk/package.json`).version as string,
+      wing_console_version: require(`@wingconsole/app/package.json`).version as string
```
Comment (eladb) [inline apps/wing/src/analytics/collectors/cli-collector.ts:27]: Maybe add some try/catch around this so that not everything fails if we can't require the module:

```js
function tryRequire(m) {
  try {
    return require(m);
  } catch {
    return undefined;
  }
}
```

File: apps/wing/src/analytics/collectors/collector.ts [hunk 28]
```diff
@@ -0,0 +1,28 @@
+import { execFile } from "child_process";
+
+export interface Data {
+  [key: string]: any;
+}
+
+export abstract class Collector {
+  abstract collect(): Promise<Data | undefined>;
+
+  async runCommand(cmd: string, args: string[]): Promise<any> {
```
Comment (eladb) [inline apps/wing/src/analytics/collectors/collector.ts:10]:
```suggestion
  protected async runCommand(cmd: string, args: string[]): Promise<any> {
```

File: apps/wing/src/analytics/storage.ts [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+   */
+  public loadConfig(): AnalyticsConfig {
+    try {
+      const fileContents = readFileSync(this.analyticsConfigFile, 'utf-8');
+      return JSON.parse(fileContents) as AnalyticsConfig;
+    } catch (error: any) {
+      // Possible issue reading analytics config file, first determine if file exists yet
+  
+      if (error.code === 'EACCES') {
+        throw new Error("Not able to access analytics config file due to permissions");
```
Comment (eladb) [inline apps/wing/src/analytics/storage.ts:148]: Is this caught somewhere? I don't think we need this to ever fail even if we can't store the file

File: apps/wing/src/analytics/storage.ts [hunk 30]
```diff
@@ -0,0 +1,202 @@
...
+  }
+
+  /**
+   * Helper method to flatten objects
+   * 
+   * @param properties The properties to flatten
+   * @param parentKey The key of the parent (defaults to "")
+   * @returns 
+   */
+  private flattenProperties(properties: {[key: string]: any}, parentKey: string = ""): {[key: string]: any} {
```
Comment (eladb) [inline apps/wing/src/analytics/storage.ts:181]: Why is this needed?

File: docs/docs/09-analytics.md [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+### What data is collected
+
+Each run of the Wing cli collects information about various aspects of the environment it is running in. Below are the analytics we gather. As well
+you can always head over to our Github and view the code that collects these metrics, as the code is open source.
+
+- The Wing CLI, SDK, and Console versions
+- The command information, including the command name, arguments, and flags
+- Operating system information, including the architecture, version, and platform
+- Node.js version
+- CI environment name (if applicable)
```
Comment (eladb) [inline docs/docs/09-analytics.md:30]: Missing anonymous git repo

**Review by eladb — APPROVED** [review]
LGTM. Some notes.

File: apps/wing/src/analytics/scripts/detached-export.ts [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+  const analytics = new Analytics({ writeKey: 'sCqPF5xSscOjJdi5Tbkqu73vfF8zkZdw'});
+  const storage = new AnalyticsStorage();
+  const event = storage.loadEvent(filePath);
+
+  if (!event) {
+    throw new Error(`No analytic event found at: ${filePath}`);
+  }
+
+  const params = {
+    anonymousId: (event.properties as any).ci_name ? 'ci_run' : storage.getAnonymousId(),
```
Comment (ekeren) [inline apps/wing/src/analytics/scripts/detached-export.ts:29]: `ci_run` is not a good value. But because I have `anonymous_repo_id` I can work the details in the dashboards. Lets revert to using `getAnonymousId`

```suggestion
    anonymousId: storage.getAnonymousId(),
```

File: apps/wing/src/analytics/storage.ts [hunk 33]
```diff
@@ -0,0 +1,202 @@
...
+   * Saves the analytics config to disk
+   * 
+   * @param config the analytics config to save to disk
+   */
+  public saveConfig(config: AnalyticsConfig) {
+    if (!existsSync(WING_HOME_DIR)) {
+      mkdirSync(WING_HOME_DIR);
+    }
+    writeFileSync(this.analyticsConfigFile, JSON.stringify(config))
+  }
```
Comment (ekeren) [inline apps/wing/src/analytics/storage.ts:172]: Maybe you should guard this with try catch ? (we might find this is a readonly fs)

File: apps/wing/src/analytics/storage.ts (hunk 30 above)
Comment (ekeren) [reply apps/wing/src/analytics/storage.ts:181]: https://github.com/winglang/wing/pull/3310#discussion_r1258312403

Comment (ekeren) [reply apps/wing/src/analytics/storage.ts:181]: some analytics consumers don't like nested property values

File: apps/wing/src/analytics/collect.ts (hunk 26 above)
Comment (ekeren) [reply apps/wing/src/analytics/collect.ts:33]: Current solution is ok (talked to Elad)

File: docs/docs/09-analytics.md [hunk 34]
```diff
@@ -0,0 +1,30 @@
+---
+title: Analytics
+id: analytics
+keywords: [Analytics, Telemetry]
+---
+
+## Analytics Collection
+
+The Wing CLI collects anonymous usage data in order to help us improve. The intent is to collect data that will help us understand how it is being used throughout the community, with hopes that this will help us to build the right features and improve the right areas.
```
Comment (staycoolcall911) [inline docs/docs/09-analytics.md:9]:
```suggestion
The Wing CLI collects anonymous usage data in order to help us improve. The intent is to collect data that will help us understand how it is being used throughout the community, with hopes that this will help us build the right features and improve the right areas.
```

File: apps/wing/src/cli.ts [hunk 35]
```diff
@@ -27,6 +39,33 @@ function actionErrorHandler(fn: (...args: any[]) => Promise<any>) {
   };
 }
 
+async function collectAnalyticsHook(cmd: Command) {
+  if (process.env.WING_DISABLE_ANALYTICS) { return; }
```
Comment (staycoolcall911) [inline apps/wing/src/cli.ts:43]: If the user opted out of anlytics collection, it would be nice to show some feedback that no analytics are sent (maybe print to stdout)

Comment (hasanaburayyan) [reply apps/wing/src/cli.ts:43]: Let me think about how best to handle output info on analytics collection. I think this might become noise, maybe its 1. if disable AND if debug then we print the message.

Ill make suggestion in subsequent pr.
